File: programmers/Lv2/프린터.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    priority_queue<int>pq;
    queue<pair<int,int>>q;
    for(int i=0;i<priorities.size();i++){
        q.push({priorities[i],i});
        pq.push(priorities[i]);
    }
    while(!q.empty()){
        int val = q.front().first;
        int idx = q.front().second;
        q.pop();
        
        if(pq.top() == val){
            pq.pop();
            answer++;
            if(location == idx)break;
            
        }else{
            q.push({val,idx});
        }
    }
    return answer;
}

File: programmers/Lv2/프린터.js
function solution(priorities, location) {
    var answer = 0;
    let pq = [];
    let q = [];
    priorities.forEach((priority,idx)=>{
        pq.push(priority);
        q.push({priority, idx});
    });
    pq = pq.sort((a,b)=>b-a);
    while(q.length !== 0){
        const {priority, idx} = q[0];
        q.shift();
        
        if(priority === pq[0]){
            pq.shift();
            answer++;
            if(idx === location)return answer;
        }else{
            q.push({priority,idx});
        }
    }
    return answer;
}
